serialized_data_interface/sdi.py

```diff
@@ -0,0 +1,377 @@
+                relation, local_versions, remote_versions
+            )
+        return local_versions_parsed[max(compatible_versions)]
+
+    def _get_entity_schema(
+        self, version: Union[int, str], entity: Union[Application, Unit]
+    ):
+        if entity is self.app:
+            return self.schemas[version].get(self.role, {}).get("app", {})
+        elif entity is self.unit:
+            return self.schemas[version].get(self.role, {}).get("unit", {})
+        elif isinstance(entity, Application):
+            return self.schemas[version].get(self.remote_role, {}).get("app", {})
+        elif isinstance(entity, Unit):
+            return self.schemas[version].get(self.remote_role, {}).get("unit", {})
+
+    def send_versions(self, relation: Relation):
+        """Send the list of supported versions to the related app.
+
+        If not the leader, does nothing.
+        """
+        if self.unit.is_leader():
+            serialized = yaml.safe_dump(list(self.versions))
+            relation.data[self.app][utils.VERSION_KEY] = serialized
+
+    def __repr__(self):
+        return (
+            f"SerializedDataInterface(charm={self.charm}, endpoint={self.endpoint}, "
+            f"schemas={self.schemas}, versions={self.versions}, role={self.role})"
+        )
+
+    @property
+    def _relations(self):
+        return [rel for rel in self.charm.model.relations[self.endpoint] if rel.app]
+
+    def get_data(self) -> Dict[Tuple[Relation, Application], Dict]:
+        """Get unwrapped data for all relations.
+
+        If any relation has an error or is unversioned, it will be surfaced immediately.
+
+        Returns a mapping of (relation, relation.app) tuples to the unwrapped data for that app.
+        """
+        data = {}
+        for relation in self._relations:
+            try:
+                rel_data = self.unwrap(relation)
+                if rel_data[relation.app]:
+                    data[(relation, relation.app)] = rel_data[relation.app]
+                if self.unit.is_leader() and rel_data[self.app]:
+                    data[(relation, self.app)] = rel_data[self.app]
+            except errors.IncompleteRelation:
+                continue
+        return data
+
+    def send_data(self, data: dict, app_name: str = None):
+        """Send data to related app(s).
+
+        If `app_name` is given, the data will only be sent to relations with that remote
+        application. Otherwise, it will be sent to all relations. Note that this means that
+        either the data given has to be compatible with any established relation versions.
+        """
+        if app_name and not any(
+            relation.app.name == app_name for relation in self._relations
+        ):
+            raise errors.InvalidAppNameError(app_name)
+        data = {self.app: data}
+        for relation in self._relations:
+            if app_name and relation.app.name != app_name:
+                continue
+            self.wrap(relation, data)
```

Comment:
yeah, the `any` i had was a leftover.  I meant what i've edited to ^^

Reply:
Hrm, still an extra loop on the non-app-name case. What about this slight modification:

```suggestion
        if not app_name:
            relations = sells._relations
        else:
            relations = [
                relation
                for relation in self._relations
                if relation.app.name == app_name
            ]
            if not relations:
                raise errors.InvalidAppNameError(app_name)
        data = {self.app: data}
        for relation in relations:
            self.wrap(relation, data)
```

Reply:
`¯\_(ツ)_/¯`
it's clear to see this isn't a gigantic big-O Problem.  We're not gonna save a lot of trees on this optimization.  Go with what reads well.  (and this suggestion ^^ has `sells` rather than `self`.  so-- don't commit it :)

Reply:
Well, yours had `self.relations` instead of `self._relations` so there! 😄 I do like it better overall, so thank you. Fixed and will commit.